pyschema/core.py

```diff
@@ -85,24 +85,29 @@ def add_record(self, schema, _bump_stack_level=False):
             Can be used as a class decorator
         """
         full_name = get_full_name(schema)
-        self._force_add(full_name, schema, _bump_stack_level)
+        has_namespace = '.' in full_name
+        self._force_add(full_name, schema, _bump_stack_level, _raise_on_existing=has_namespace)
         if '.' in full_name and schema.__name__ not in self._schema_map:
             self._force_add(schema.__name__, schema, _bump_stack_level)
         return schema
 
-    def _force_add(self, used_name, schema, _bump_stack_level=False):
+    def _force_add(self, used_name, schema, _bump_stack_level=False, _raise_on_existing=False):
         existing = self._schema_map.get(used_name, None)
         if existing:
-            full_name = get_full_name(schema)
-            explanation = "(actually {0})".format() if full_name != used_name else ""
-
-            warnings.warn(
-                "{used_name}{explanation}: old definition in {prev_module} replaced by definition in {new_module}"
-                .format(used_name=used_name,
-                        explanation=explanation,
-                        prev_module=existing.__module__,
-                        new_module=schema.__module__),
-                stacklevel=4 if _bump_stack_level else 3)
+            if existing != schema and _raise_on_existing:
+                raise ValueError('Attempted to add a new schema replacing an existing one. '
```

Comment:
Couldn't the `existing != schema` check be on the outer if, since there is technically no collision if it's just the same schema reference being re-added